compiler/rustc_parse_format/src/lib.rs

```diff
@@ -823,6 +829,27 @@ impl<'a> Parser<'a> {
         if found { Some(cur) } else { None }
     }
 
+    fn suggest_format(&mut self) {
+        if let (Some(pos), Some(_)) = (self.consume_pos('?'), self.consume_pos(':')) {
+            let word = self.word();
+            let _end = self.current_pos();
+            let pos = self.to_span_index(pos);
+            self.errors.insert(
+                0,
+                ParseError {
+                    description: "expected format parameter to occur after `:`".to_owned(),
+                    note: Some(
+                        format!("`?` comes after `:`, try `{}:{}` instead", word, "?").to_owned(),
+                    ),
+                    label: "expected `?` to occur after `:`".to_owned(),
+                    span: pos.to(pos),
+                    secondary_label: None,
+                    should_be_replaced_with_positional_argument: false,
+                },
+            );
```

Comment:
What's the behavior in the face of `{?:?}`, `{??}`, `{?;foo}`, `{?:#?}`?

Reply:
That's a very good question :) 

`{?:?}` results in: 
```
error: invalid format string: expected format parameter to occur after `:`
 --> test_three.rs:4:16
  |
4 |     println!("{?:?}", value);
  |                ^ expected `?` to occur after `:` in format string
  |
  = note: `?` comes after `:`, try `:?` instead
  ```
  
  `{??}` results in: 
  ```
error: invalid format string: expected `'}'`, found `'?'`
 --> test_three.rs:5:16
  |
5 |     println!("{??}", value);
  |               -^ expected `}` in format string
  |               |
  |               because of this opening brace
  |
  = note: if you intended to print `{`, you can escape it using `{{`
```

`{?;value}`results in:
```
error: invalid format string: expected `'}'`, found `'?'`
 --> test_three.rs:6:16
  |
6 |     println!("{?;value}");
  |               -^ expected `}` in format string
  |               |
  |               because of this opening brace
  |
  = note: if you intended to print `{`, you can escape it using `{{`
```

And finally, `{?:#?}` results in:
```
error: invalid format string: expected format parameter to occur after `:`
 --> test_three.rs:7:16
  |
7 |     println!("{?:#?}", value);
  |                ^ expected `?` to occur after `:` in format string
  |
  = note: `?` comes after `:`, try `:?` instead
```

I honestly don't dislike these answers. The places where we are fairly certain what the user is doing, we can provide a good hint and the places where we can't be totally sure we fall back to the standard one there is there already. What do you think?

Reply:
Lets add these as tests.

Reply:
I don't know why I didn't think of that. Of course, we should add those. I have done so now. @estebank (I don't know if the @ mention is necessary).